src/criticality/BoratedWater.C

```diff
@@ -0,0 +1,223 @@
+
+registerMooseObject("CardinalApp", BoratedWater);
+
+InputParameters
+BoratedWater::validParams()
+{
+  auto params = OpenMCMaterialSearch::validParams();
+  params.addParam<std::vector<std::string>>(
+      "absent_nuclides",
+      "Natural nuclides of hydrogen, oxygen, or boron which are missing from your cross section "
+      "library; some cross section libraries do not have entries for O17 and O18. If your library "
+      "does not have these nuclides you will get an error from this object trying to add them. For "
+      "these missing nuclides, specify them here and their abundance will be applied to the main "
+      "isotope of each element.");
+  params.addClassDescription(
+      "Searches for criticality using natural boron ppm in water in units of weight ppm");
+  return params;
+}
+
+BoratedWater::BoratedWater(const InputParameters & parameters) : OpenMCMaterialSearch(parameters)
+{
+  // apply additional checks on the minimum and maximum
+  if (_minimum < 0)
+    paramError("minimum",
+               "The 'minimum' boron ppm (" + std::to_string(_minimum) + ") must be positive!");
+  if (_maximum > 50000)
+    paramError("maximum",
+               "The borated water composition is computed using a dilute species approximation. "
+               "Results will not be accurate if the boron species is no longer dilute, which we "
+               "take as 5\% weight concentration or higher. Please decrease 'maximum' (" +
+                   std::to_string(_maximum) + ") to an upper limit which is in the dilute regime.");
+
+  // We take the provided material, retaining its density, but then overwriting
+  // any nuclides there to be H2O + boron weight ppm and add the S(a,b) tables
+  // for hydrogen. Therefore, we perform a check here if any nuclides in the
+  // material are *not* those we are going to add back when we wipe the material,
+  // to notify the user they probably provided the wrong material ID or they need
+  // to use a more general material modification object that allows full control
+  // over additional nuclides (e.g., if their water includes corrosion products
+  // that they do want to be there).
+
+  const int * nuclides;
+  const double * densities;
+  int n;
+  int err = openmc_material_get_densities(_material_index, &nuclides, &densities, &n);
+  catchOpenMCError(err, "get nuclide densities from material " + std::to_string(_material_id));
+
+  // get all the natural isotopes of H, O, B
+  _hydrogen_natural = NuclearData::Nuclide::getAbundances("H");
+  _boron_natural = NuclearData::Nuclide::getAbundances("B");
+  _oxygen_natural = NuclearData::Nuclide::getAbundances("O");
+
+  std::vector<std::string> allowable;
+  for (const auto & i : _hydrogen_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _boron_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _oxygen_natural)
+    allowable.push_back(i.first);
+
+  if (isParamValid("absent_nuclides"))
+  {
```

Comment:
It seems useful to me for this block to be in its own private function.  It can help the overall readability by reducing the cognitive burden of this function.